cirq/ops/fourier_transform.py

```diff
@@ -0,0 +1,187 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+
+import numpy as np
+import sympy
+
+import cirq
+from cirq import value, _compat
+from cirq.ops import raw_types
+
+
+@value.value_equality
+class QuantumFourierTransformGate(raw_types.Gate):
+    """Switches from the computational basis to the frequency basis."""
+
+    def __init__(self, num_qubits: int, *, without_reverse: bool = False):
+        """
+        Args:
+            num_qubits: The number of qubits the gate applies to.
+            without_reverse: Whether or not to include the swaps at the end
+                of the circuit decomposition that reverse the order of the
+                qubits. These are technically necessary in order to perform the
+                correct effect, but can almost always be optimized away by just
+                performing later operations on different qubits.
+        """
+        self._num_qubits = num_qubits
+        self._without_reverse = without_reverse
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'num_qubits': self._num_qubits,
+            'without_reverse': self._without_reverse
+        }
+
+    def _value_equality_values_(self):
+        return self._num_qubits, self._without_reverse
+
+    def num_qubits(self) -> int:
+        return self._num_qubits
+
+    def _decompose_(self, qubits):
+        if len(qubits) == 0:
+            return
+        yield cirq.H(qubits[0])
+        for i in range(1, len(qubits)):
+            yield PhaseGradientGate(
+                num_qubits=i,
+                exponent=0.5).on(*qubits[:i][::-1]).controlled_by(qubits[i])
+            yield cirq.H(qubits[i])
+        if not self._without_reverse:
+            for i in range(len(qubits) // 2):
+                yield cirq.SWAP(qubits[i], qubits[-i - 1])
+
+    def _has_unitary_(self):
+        return True
+
+    def __str__(self):
+        return 'QFT[norev]' if self._without_reverse else 'QFT'
+
+    def __repr__(self):
+        return ('cirq.QuantumFourierTransformGate(num_qubits={!r}, '
+                'without_reverse={!r})'.format(self._num_qubits,
+                                               self._without_reverse))
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'):
+        return cirq.CircuitDiagramInfo(
+            wire_symbols=(str(self),) +
+            tuple(f'#{k+1}' for k in range(1, self._num_qubits)),
+            exponent_qubit_index=0)
+
+
+@value.value_equality
+class PhaseGradientGate(raw_types.Gate):
```

Comment:
Discussed IRL and decided this wouldn't work because generating and iterating over the eigenprojectors scales poorly with n. EigenGate works best for small gates, but this gate has arbitrary size.